File: packages/protocol/test/canonical.test.ts
/**
 * The canonical encoder pins the byte-level choices where Node and Go would
 * otherwise disagree. `connectorv1/canonical_test.go` asserts the same table
 * against the Go implementation.
 */

import assert from "node:assert/strict";
import test from "node:test";

import {
  byteLength,
  CanonicalEncodeError,
  jsonBoolean,
  jsonInteger,
  jsonNumber,
  jsonString,
} from "../src/canonical.ts";

test("numbers are formatted by the ECMAScript algorithm", () => {
  const cases: [number, string][] = [
    [0, "0"],
    [-0, "0"],
    [1, "1"],
    [-1, "-1"],
    [0.42, "0.42"],
    [1.5, "1.5"],
    [100, "100"],
    [8132, "8132"],
    [8200000000, "8200000000"],
    [9007199254740991, "9007199254740991"],
    [0.000001, "0.000001"],
    [1e-7, "1e-7"],
    [1.25e-8, "1.25e-8"],
    [1e21, "1e+21"],
    [1.5e22, "1.5e+22"],
    [1e20, "100000000000000000000"],
    [0.1, "0.1"],
    [1 / 3, "0.3333333333333333"],
    [1024, "1024"],
  ];
  for (const [value, expected] of cases) {
    assert.equal(jsonNumber(value), expected, `jsonNumber(${String(value)})`);
  }
});

test("strings are escaped the way the Go writer escapes them", () => {
  const cases: [string, string][] = [
    ["plain", '"plain"'],
    ['quote"inside', '"quote\\"inside"'],
    ["back\\slash", '"back\\\\slash"'],
    ["tab\there", '"tab\\there"'],
    ["line\nbreak", '"line\\nbreak"'],
    ["bell", '"bell\\u0007"'],
    ["<tag> & 'amp'", "\"<tag> & 'amp'\""],
    ["日本語", '"日本語"'],
    ["\u{1f680}", '"\u{1f680}"'],
    ["line separator", '"line\\u2028separator"'],
    ["paragraph separator", '"paragraph\\u2029separator"'],
  ];
  for (const [value, expected] of cases) {
    assert.equal(jsonString(value), expected, `jsonString(${JSON.stringify(value)})`);
  }
});

test("non-finite and unsafe numbers are refused", () => {
  assert.throws(() => jsonNumber(Number.POSITIVE_INFINITY), CanonicalEncodeError);
  assert.throws(() => jsonNumber(Number.NaN), CanonicalEncodeError);
  assert.throws(() => jsonInteger(1.5), CanonicalEncodeError);
  assert.throws(() => jsonInteger(Number.MAX_SAFE_INTEGER + 2), CanonicalEncodeError);
});

test("booleans and byte lengths", () => {
  assert.equal(jsonBoolean(true), "true");
  assert.equal(jsonBoolean(false), "false");
  assert.equal(byteLength("abc"), 3);
  assert.equal(byteLength("é"), 2, "bounds are byte bounds, not character counts");
  assert.equal(byteLength(new Uint8Array([1, 2, 3, 4])), 4);
});

File: packages/protocol/connectorv1/canonical_test.go
package connectorv1

import (
	"strings"
	"testing"
)

// The canonical writer must format numbers and strings exactly as the
// TypeScript encoder does, or the two languages cannot produce byte-identical
// frames. The expectations below are the output of ECMAScript's
// String(value) and JSON.stringify, which the TypeScript encoder uses directly.

func TestFormatECMANumberMatchesECMAScript(t *testing.T) {
	cases := []struct {
		value    float64
		expected string
	}{
		{0, "0"},
		{-0, "0"},
		{1, "1"},
		{-1, "-1"},
		{0.42, "0.42"},
		{1.5, "1.5"},
		{100, "100"},
		{8132, "8132"},
		{8200000000, "8200000000"},
		{9007199254740991, "9007199254740991"},
		{0.000001, "0.000001"},
		{1e-7, "1e-7"},
		{1.25e-8, "1.25e-8"},
		{1e21, "1e+21"},
		{1.5e22, "1.5e+22"},
		{1e20, "100000000000000000000"},
		{0.1, "0.1"},
		{1.0 / 3.0, "0.3333333333333333"},
		{1024, "1024"},
	}
	for _, testCase := range cases {
		if formatted := formatECMANumber(testCase.value); formatted != testCase.expected {
			t.Errorf("formatECMANumber(%v) = %q, want %q", testCase.value, formatted, testCase.expected)
		}
	}
}

func TestAppendJSONStringEscapesLikeECMAScript(t *testing.T) {
	cases := []struct {
		value    string
		expected string
	}{
		{"plain", `"plain"`},
		{`quote"inside`, `"quote\"inside"`},
		{`back\slash`, `"back\\slash"`},
		{"tab\there", `"tab\there"`},
		{"line\nbreak", `"line\nbreak"`},
		{"bell\u0007", `"bell\u0007"`},
		// HTML-significant characters are not escaped, matching JSON.stringify
		// and unlike Go's encoding/json default.
		{"<tag> & 'amp'", `"<tag> & 'amp'"`},
		{"日本語", `"日本語"`},
		{"\U0001f680", "\"\U0001f680\""},
		// U+2028 and U+2029 are escaped for parity with the TypeScript encoder.
		{"line\u2028separator", `"line\u2028separator"`},
		{"paragraph\u2029separator", `"paragraph\u2029separator"`},
	}
	for _, testCase := range cases {
		if formatted := string(appendJSONString(nil, testCase.value)); formatted != testCase.expected {
			t.Errorf("appendJSONString(%q) = %s, want %s", testCase.value, formatted, testCase.expected)
		}
	}
}

func TestCanonicalWriterReportsNonFiniteNumbers(t *testing.T) {
	var writer canonicalWriter
	writer.beginObject()
	writer.key("load")
	writer.number(positiveInfinity())
	writer.endObject()
	if _, err := writer.result(); err == nil {
		t.Fatal("expected a non-finite number to be reported")
	}
}

func positiveInfinity() float64 {
	large := 1e308
	return large * 10
}

// Every generated pattern must compile under RE2 as well as under the
// JavaScript engine. Compilation happens at package initialisation, so reaching
// this test at all proves it; the assertion guards against an empty set.
func TestGeneratedPatternsCompile(t *testing.T) {
	if pattern0 == nil {
		t.Fatal("no generated patterns were compiled")
	}
	if !pattern0.MatchString("con_01JQ8ZR3M4T5V6W7X8Y9Z0A1C2") {
		t.Fatal("the identifier pattern rejects a conventional identifier")
	}
	if pattern0.MatchString("con id") {
		t.Fatal("the identifier pattern accepts whitespace")
	}
	if strings.Contains(pattern0.String(), "con_") {
		t.Fatal("the identifier pattern must not constrain the prefix")
	}
}
